### PR TITLE
Image expression

### PR DESCRIPTION
**Pull request by ryanhamley**
## Launch Checklist

This PR implements a new `Image` expression type and `image` operator as discussed in https://github.com/mapbox/mapbox-gl-js/issues/8052# particularly https://github.com/mapbox/mapbox-gl-js/issues/8052#issuecomment-511622996

 - [x] briefly describe the changes in this PR
    - adds an `image` expression operator to allow for determining an image's availability before requesting it
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [ ] post benchmark scores
 - [x] manually test the debug page
 - [x] tagged `@mapbox/studio` and/or `@mapbox/maps-design` if this PR includes style spec changes
 - [x] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
    - will need a native port @mapbox/gl-native 

**Changelog**
<changelog>Add `image` expression operator which determines if a requested image is available in the map's style or not</changelog>

### CODE REVIEW

File: src/style-spec/expression/definitions/image.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        return null;
+    }
+
+    eachChild(fn: (Expression) => void) {
+        fn(this.input);
+    }
+
+    possibleOutputs() {
+        // Technically the combinatoric set of all children
```
[the remark this answers is not in the document]
Comment (ryanhamley) [reply src/style-spec/expression/definitions/image.js]: I'm not sure what `possibleOutputs` is for. It haven't needed it for this. This comment was copied over from elsewhere. I can update the comment.

File: src/render/image_atlas.js [hunk 2]
```diff
@@ -140,4 +140,3 @@ export default class ImageAtlas {
 
 register('ImagePosition', ImagePosition);
 register('ImageAtlas', ImageAtlas);
-
```
Comment (asheemmamoowala) [inline src/render/image_atlas.js]: nit: no change needed

File: src/style-spec/expression/definitions/image.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        this.input = input;
+    }
+
+    static parse(args: $ReadOnlyArray<mixed>, context: ParsingContext): ?Expression {
+        if (args.length !== 2) {
+            return context.error(`Expected two arguments.`);
+        }
+
+        const name = context.parse(args[1], 1, StringType);
+        if (!name) return context.error(`No icon name provided.`);
```
Comment (asheemmamoowala) [inline src/style-spec/expression/definitions/image.js]: nit: Error should refer to `image` not `icon`

File: src/style-spec/reference/v8.json [hunk 4]
```diff
@@ -2804,6 +2804,15 @@
           }
         }
       },
+      "image": {
+        "doc": "Returns `image` type for use in `icon-image` and `*-pattern` entries. If set, the `image` argument will check that the requested image exists in the style and will return either the resolved image name or `null`, depending on whether or not the image is currently in the style. This validation process is synchronous and requires the image to have been added to the style before requesting it in the `image` argument.",
```
Comment (asheemmamoowala) [inline src/style-spec/reference/v8.json]:
```suggestion
        "doc": "Returns an `image` type for use in `icon-image` and `*-pattern` entries. This expression validates that a named image is available in the style, returning `null` if it is not found. This validation process is synchronous and requires the image to have been added to the style before requesting it in the `image` argument.",
```

File: src/render/draw_background.js [hunk 5]
```diff
@@ -25,7 +25,7 @@ function drawBackground(painter: Painter, sourceCache: SourceCache, layer: Backg
     const transform = painter.transform;
     const tileSize = transform.tileSize;
     const image = layer.paint.get('background-pattern');
-    if (painter.isPatternMissing(image)) return;
+    if (painter.isPatternMissing((image: any))) return;
```
Comment (asheemmamoowala) [inline src/render/draw_background.js]: Is Flow complaining because the expected type in `isPatternMissing` is `CrossFaded<string>` but `layer.paint.get` is now returning a `CrossFaded<Image>`. If so would it make sense to update `isPatternMissing` and undo the changes in `draw_line`, `draw_fill`, and `draw_fill_extrusion`

Comment (ryanhamley) [reply src/render/draw_background.js]: `isPatternMissing` expects `CrossFaded<string>`. I can change that to `CrossFaded<Image>` and then in `isPatternMissing` call `image.to.toString()` and it works well and gets rid of the Flow error.

If I try that same approach in the other `draw_*` files though, I get a weird Flow error. 

```
        if (constantPattern && tile.imageAtlas) {
            const posTo = tile.imageAtlas.patternPositions[constantPattern.to.toString()];
            const posFrom = tile.imageAtlas.patternPositions[constantPattern.from.toString()];
            if (posTo && posFrom) programConfiguration.setConstantPatternPositions(posTo, posFrom);
        }
```

gets rid of the `Image is incompatible with string` error (patternPositions expects a string key) but Flow then complains that it `Cannot get tile.imageAtlas.patternPositions because property patternPositions is missing in null or undefined` which doesn't make sense since we're already checking that `tile.imageAtlas` isn't null or undefined.

Comment (ryanhamley) [reply src/render/draw_background.js]: Actually, I'm wrong. I had forgotten to remove the cast to `any` in `draw_background`. If I do that and update `isPatternMissing` I get this error:

`Cannot call painter.isPatternMissing with image bound to image because Image [1]
is incompatible with Image [2] in property from.`

which doesn't make sense since `to` and `from` are just defined as the type passed to `CrossFaded`

https://github.com/mapbox/mapbox-gl-js/blob/3d6d2b440c89199dfc38f0a2b109588f1f264ac8/src/style/properties.js#L28-L31

File: src/render/image_atlas.js (hunk 2 above)
Comment (ryanhamley) [reply src/render/image_atlas.js]: This was automatically added by my editor. It's pretty standard to end files with a new line isn't it? Most of our other files have a trailing new line.

File: src/render/draw_background.js (hunk 5 above)
Comment (ryanhamley) [reply src/render/draw_background.js]: @ahk helped me fix the casts to `any` in `draw_{fill/fill-extrusion/line}` but we haven't come up with a way around this cast yet. We're still trying.

Comment (ryanhamley) [reply src/render/draw_background.js]: Turns out that type `Image` accidentally shadows the Javascript constructor `Image` leading to a very difficult to debug Flow error. To avoid this, I'm going to change the `Image` type to `ResolvedImage`

Comment (ryanhamley) [reply src/render/draw_background.js]: All casts to `any` have been removed. The Flow typing should be solid now

**Comment by ryanhamley** [comment]
>Question, do you plan to add support for default image that is not in a sprite sheet? Or, would it be handled via onStyleImageMissing callback?

I'm adding an additional step in a follow-up PR to determine the first requested image in an expression (reading from left to right and assuming that the first image requested is the "most desirable" option) which evaluates to `null` and returning that image's name so that `styleimagemissing` will always work as expected.

>Another question is that, what happens when developer removes image from a sprite sheet while worker is parsing tile? It looks like availableImages and images that are in sprite sheet could be out-of-sync and image operator would not fall-back to default value (e.g., when used in coalesce).

I think this is analogous to what happens if you asynchronously load an image with `addImage`. Before this PR, asynchronously loading an image will not cause it to appear in a tile that had requested the image previous to it being loaded. This PR does not change that behavior because it would require reloading tiles every time an image is added or removed, which likely would be distracting and probably a performance hit. I think the best way to deal with both of these situations is still an open question.
